### PR TITLE
Imports & WdlNamespacing

### PR DESCRIPTION
**Pull request by ruchim**

### CODE REVIEW

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 1]
```diff
@@ -222,7 +227,7 @@ object WdlNamespace {
 
     def getScope(scopeAst: Ast, parent: Option[Scope]): Scope = {
       val scope = scopeAst.getName match {
-        case AstNodeName.Call => Call(scopeAst, namespaces, tasks, wdlSyntaxErrorFormatter)
+        case AstNodeName.Call => Call(scopeAst, namespaces, topLevelTasks, wdlSyntaxErrorFormatter)
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlNamespace.scala]: Am I missing something? Looks like the second `case AstNodeName.Call` will never hit if this one catches everything?

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 2]
```diff
@@ -259,9 +264,10 @@ object WdlNamespace {
         case AstNodeName.Task => getTaskInputsOutputs(scopeAst)
         case AstNodeName.Declaration | AstNodeName.Output => Seq.empty[Scope]
         case AstNodeName.Call =>
-          val referencedTask = findTask(scopeAst.getAttribute("task").sourceString, namespaces, tasks)
+          val referencedTask = findTask(scopeAst.getAttribute("task").sourceString, namespaces, topLevelTasks)
           referencedTask match {
             case Some(task) => getScopeAsts(task.ast, "declarations").map(d => getScope(d, scope))
+            // TODO: sfrazer: uh oh... syntax errorƒ
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlNamespace.scala]: A ghost at the feast?

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 3]
```diff
@@ -270,16 +276,16 @@ object WdlNamespace {
...
-    val children = tasks ++ namespaces ++ nonTaskScopes.map(ast => getScope(ast, parent = None))
+    val children = topLevelTasks ++ namespaces ++ nonTaskScopes.map(ast => getScope(ast, parent = None))
 
     val namespace = children.collect({ case w: Workflow => w }) match {
-      case Nil => WdlNamespaceWithoutWorkflow(namespaceName, imports, namespaces, tasks, terminalMap, ast)
-      case Seq(workflow) => WdlNamespaceWithWorkflow(ast, workflow, namespaceName, imports, namespaces, tasks, terminalMap, wdlSyntaxErrorFormatter)
-      case _ => throw new SyntaxError(wdlSyntaxErrorFormatter.tooManyWorkflows(ast.findAsts(AstNodeName.Workflow).asJava))
+      case Nil => WdlNamespaceWithoutWorkflow(namespaceName, imports, namespaces, topLevelTasks, terminalMap, ast)
+      case Seq(workflow) => WdlNamespaceWithWorkflow(ast, workflow, namespaceName, imports, namespaces, topLevelTasks, terminalMap, wdlSyntaxErrorFormatter)
+      case x => throw new SyntaxError(wdlSyntaxErrorFormatter.tooManyWorkflows(ast.findAsts(AstNodeName.Workflow).asJava))
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlNamespace.scala]: Unused `x`. Could stay as `_` although I'd prefer to name it and use it!

File: src/test/cases/command_parameters/ancestry.json [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  "param_test.f": ["param_test", ""],
+  "wf": [""],
+  "wf.param_test": ["wf", ""],
+  "wf.param_test.a": ["wf.param_test", "wf", ""],
+  "wf.param_test.b": ["wf.param_test", "wf", ""],
+  "wf.param_test.c": ["wf.param_test", "wf", ""],
+  "wf.param_test.d": ["wf.param_test", "wf", ""],
+  "wf.param_test.e": ["wf.param_test", "wf", ""],
+  "wf.param_test.f": ["wf.param_test", "wf", ""]
+}
```
Comment (cjllanwarne) [inline src/test/cases/command_parameters/ancestry.json]: Newline. Is this your change of Scott's?

Comment (cjllanwarne) [inline src/test/cases/command_parameters/ancestry.json]: Also all the other test files :-/

File: src/test/scala/wdl4s/WorkflowImportsSpec.scala [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    |  call ${noExtension(printNumsWdlImportFile)}.multilingualEcho.inItalian
+    |
+    |}
+    |""".stripMargin
+
+  val wdlWithImports = imports + primaryWorkflow
+
+  val namespace = WdlNamespaceWithWorkflow.load(wdlWithImports, wdlDirectory)
+
+  "WDL file with imports" should "Have 0 tasks (3 tasks are in separate namespace)" in {
```
Comment (cjllanwarne) [inline src/test/scala/wdl4s/WorkflowImportsSpec.scala]: ToL: I'd rather move all the test cases to the top (i.e. above all the sample WDL). Also a lot of the name sample WDLs seem unused?

Comment (cjllanwarne) [inline src/test/scala/wdl4s/WorkflowImportsSpec.scala]: Is there scope for more tests here, e.g. 
- `it should "import a WDL file (with alias) and be able to reference its tasks by FQN"`
- `it should "import a WDL file (with no alias) and be able to reference its tasks by FQN"`
- `it should "import two WDL file (with clashing task names) and be able to reference all tasks by FQN"`
- `it should "Not be able to import WDLs with clashing namespaces"`
- etc...

**Comment by cjllanwarne** [comment]
Test failures look genuine... :'(

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 6]
```diff
@@ -143,76 +146,78 @@ case class WdlNamespaceWithWorkflow(importedAs: Option[String],
...
-      * For import statements with an 'as' clause, this translates to a sub-namespace
-      */
-    val namespaces: Seq[WdlNamespace] = for {
-      imp <- imports if imp.namespaceName.isDefined
-      source = importResolver(imp.uri)
-      if source.length > 0
-    } yield WdlNamespace.load(source, imp.uri, importResolver, imp.namespaceName)
+    def tryResolve(str: String): Option[WdlSource] = {
+      importResolvers map { resolver =>
+        Try(resolver(str))
```
Comment (geoffjentry) [inline src/main/scala/wdl4s/WdlNamespace.scala]: "Does X more naturally a Future"?

File: src/main/scala/wdl4s/Import.scala [hunk 7]
```diff
@@ -13,5 +14,6 @@ object Import {
 
 // FIXME: I dislike dragging the AST along but it's necessary for "compile" time error syntax highlighting, argh
 case class Import(uri: String, namespaceTerminal: Option[Terminal]) {
-  val namespaceName: Option[String] = namespaceTerminal.map(_.sourceString)
+  val namespaceName: String = namespaceTerminal.map(_.sourceString).getOrElse(File(uri).nameWithoutExtension)
+  val usingDefaultNamespace = namespaceTerminal.map(_.sourceString).isEmpty
```
Comment (ruchim) [inline src/main/scala/wdl4s/Import.scala]: great catch, been removed.

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 8]
```diff
@@ -143,76 +146,78 @@ case class WdlNamespaceWithWorkflow(importedAs: Option[String],
...
+    // Would like to rename this to something more appropriate
+    val topLevelAsts = ast.getAttribute("body").astListAsVector.collect({ case a: Ast => a })
+
     /**
-      * All `task` definitions, including ones from import statements with no 'as' clause
+      * All `task` definitions of primary workflow.
       */
-    val tasks: Seq[Task] = for {
-      taskAst <- topLevelScopeAsts if taskAst.getName == AstNodeName.Task
+    val topLevelTasks: Seq[Task] = for {
```
Comment (ruchim) [inline src/main/scala/wdl4s/WdlNamespace.scala]: Imports are still recursive, I believe I need this for the separation of tasks in the primary workflow vs imported workflows. In order to find a referenced task in a workflow with imports, the findTask() method uses the primary workflow tasks (topLevelTasks) when searching for calls without the "." Its possible that it can be restructured with less code?

File: src/main/scala/wdl4s/WdlNamespace.scala (hunk 3 above)
Comment (ruchim) [inline src/main/scala/wdl4s/WdlNamespace.scala]: I will utilize the power of the x variable

File: src/main/scala/wdl4s/WdlNamespace.scala (hunk 1 above)
Comment (ruchim) [inline src/main/scala/wdl4s/WdlNamespace.scala]: what's the second case you're referring to? (other than the one on line 230)

File: src/test/cases/command_parameters/ancestry.json (hunk 4 above)
Comment (ruchim) [inline src/test/cases/command_parameters/ancestry.json]: The new PR now no longer has Scott's changes incorporated. Had to make a separate PR.

File: src/test/scala/wdl4s/WorkflowImportsSpec.scala (hunk 5 above)
Comment (ruchim) [inline src/test/scala/wdl4s/WorkflowImportsSpec.scala]: Since the tests rely on creation of temp directories and files,  I was unable to move to the tests to the top. I didn't find anything unused, but let me know if you spot something specific? @mcovarr maybe its worth adding the compiler flag related to unused vals/methods in WDL4s as well, I've really found it useful.

**Comment by ruchim** [comment]
PR ready for re-review in #46
